boot/bootstate20.go

```diff
@@ -810,3 +921,69 @@ func (bsmark *bootState20MarkSuccessful) commit() error {
 
 	return nil
 }
+
+// genericInitramfsSelectSnap will run the logic to choose which snap should be
+// mounted during the initramfs using the given bootState and the expected try
+// status. The try status is needed because during the initramfs we will have
+// different statuses for kernel vs base snaps, where base snap is expected to
+// be in "try" mode, but kernel is expected to be in "trying" mode. It returns
+// the first and second choice for what snaps to mount. If there is a second
+// snap, then that snap is the fallback or non-trying snap and the first snap is
+// the try snap.
+func genericInitramfsSelectSnap(bs bootState, expectedTryStatus, typeString string) (
+	firstChoice, secondChoice snap.PlaceInfo,
+	err error,
+) {
+	curSnap, trySnap, snapTryStatus, err := bs.revisions()
+
+	if err != nil && !isTrySnapError(err) {
+		// we have no fallback snap!
+		return nil, nil, fmt.Errorf("fallback %s snap unusable: %v", typeString, err)
+	}
+
+	// check that the current snap actually exists
+	file := curSnap.Filename()
+	snapPath := filepath.Join(dirs.SnapBlobDirUnder(InitramfsWritableDir), file)
+	if !osutil.FileExists(snapPath) {
+		// somehow the boot snap doesn't exist in ubuntu-data
+		// for a kernel, this could happen if we have some bug where ubuntu-boot
+		// isn't properly updated and never changes, but snapd thinks it was
+		// updated and eventually snapd garbage collects old revisions of
+		// the kernel snap as it is "refreshed"
+		// for a base, this could happen if the modeenv is manipulated
+		// out-of-band from snapd
+		return nil, nil, fmt.Errorf("%s snap %q does not exist on ubuntu-data", typeString, file)
+	}
+
+	if err != nil && isTrySnapError(err) {
+		// just log that we had issues with the try snap and continue with
```

Comment:
we don't seem to exercise this nor the places that use newTrySnapError

Reply:
The check in `isTrySnapErr()` is wrong. I've pushed a patch fixing the check and extending coverage.

Reply:
thanks @bboozzoo